src/stan_math_backend/Cpp.ml

```diff
@@ -0,0 +1,724 @@
+    | Constructor (t, es) ->
+        pf ppf "@[<hov 2>%a(%a)@]" pp_type_ t (list ~sep:comma pp_expr) es
+    | AllocNew (t, es) ->
+        pf ppf "@[<hov 2>new %a(%a)@]" pp_type_ t (list ~sep:comma pp_expr) es
+    | OperatorNew (ptr, t, es) ->
+        pf ppf "@[<hov 2>new %a%a(%a)@]"
+          (trailing_space (parens string))
+          ("&" ^ ptr) pp_type_ t (list ~sep:comma pp_expr) es
+    | ArrayLiteral es -> pf ppf "{%a}" (list ~sep:comma pp_expr) es
+    | InitializerExpr (t, es) ->
+        pf ppf "@[<2>%a{%a}@]" pp_type_ t (list ~sep:comma pp_expr) es
+    | StreamInsertion (e, es) ->
+        pf ppf "%a <<@[@ %a@]" pp_expr e (list ~sep:comma pp_expr) es
+    | FunCall (fn, tys, es) ->
+        pf ppf "@[<hov 2>%s%a(@,%a@])" fn maybe_templates tys
+          (list ~sep:comma pp_expr) es
+    | MethodCall (e, fn, tys, es) ->
+        pf ppf "@[<hov 2>%a.%s%a(%a)@]" pp_expr e fn maybe_templates tys
+          (list ~sep:comma pp_expr) es
+    | StaticMethodCall (ty, fn, tys, es) ->
+        pf ppf "@[<hov 2>%a::%s%a(%a)@]" pp_type_ ty fn maybe_templates tys
+          (list ~sep:comma pp_expr) es
+    | TernaryIf (e1, e2, e3) ->
+        pf ppf "%a ? %a : %a" pp_expr e1 pp_expr e2 pp_expr e3
+    | Index (e1, e2) -> pf ppf "%a[%a]" pp_expr e1 pp_expr e2
+    | Assign (e1, e2) -> pf ppf "%a = %a" pp_expr e1 pp_expr e2
+    | PMinus e -> pf ppf "-%a" pp_expr e
+    | Increment e -> pf ppf "++%a" pp_expr e
+    | BinOp (e1, op, e2) ->
+        pf ppf "%a@ %a@ %a" pp_expr e1 pp_operator op pp_expr e2
+
+  let pp_variable_defn ppf {static; constexpr; type_; name; init} =
+    let pp_init ppf init =
+      match init with
+      | Uninitialized -> ()
+      | Assignment e -> pf ppf " =@ %a" pp_expr e
+      | Construction es -> pf ppf "(%a)" (list ~sep:comma pp_expr) es
+      | InitalizerList es ->
+          pf ppf "{@[<hov 1>%a@]}" (list ~sep:comma pp_expr) es in
+    let static = if static then "static " else "" in
+    let constexpr = if constexpr then "constexpr " else "" in
+    pf ppf "@[<2>%s%s%a %s%a@]" static constexpr pp_type_ type_ name pp_init
+      init
+
+  let rec pp_stmt ppf s =
+    match s with
+    | Expression e -> pf ppf "@[<2>%a;@]" pp_expr e
+    | Return e -> pf ppf "return %a;" (option (box pp_expr)) e
+    | Throw e -> pf ppf "throw %a;" pp_expr e
+    | Break -> string ppf "break;"
+    | Continue -> string ppf "continue;"
+    | Semicolon -> string ppf ";"
+    | VariableDefn vd -> pf ppf "%a;" pp_variable_defn vd
+    | For (init, cond, incr, Block stmts) ->
+        (* When we know a block is here, we can do better pretty-printing*)
+        pf ppf "@[<v 2>for(@[<hov>%a; %a; %a@]) {@ @[<v>%a@]@]@,}"
+          pp_variable_defn init pp_expr cond pp_expr incr
+          (list ~sep:cut pp_stmt) stmts
+    | For (init, cond, incr, s) ->
+        pf ppf "@[<v 2>for(@[<hov>%a; %a; %a@])@ @[%a@]@]" pp_variable_defn init
+          pp_expr cond pp_expr incr pp_stmt s
```

Comment:
That better pretty-printing for blocks could be refactored into a helper function (not sure what to call it)
```ocaml
let pp_with_block pp ppf = function
  | Block stmts ->
      pf ppf "@[<v 2>%a {@ @[<v>%a@]@]@,}" pp () (list pp_stmt) stmts
  | stmt -> pf ppf "@[<v 2>%a@ @[<v>%a@]@]" pp () pp_stmt stmt
```
and then you can simplify this
```suggestion
    | For (init, cond, incr, s) ->
        let pp ppf () =
          pf ppf "for (@[<hov>%a; %a; %a@])" pp_variable_defn init pp_expr cond
            pp_expr incr in
        pp_with_block pp ppf s
```
(I added a space to `for(` as that looks a bit cleaner.)
The same helper can be used for `ForEach`, `While`, and `IfElse` (and those also need a space between the keyword and paren.)

Reply:
This took some tweaking to get right for `if/else` but looks really nice in the end!